### PR TITLE
bpo-34284: Nonsensical exception message when calling `__new__` on non-instantiable objects

### PR DESCRIPTION
**Pull request by VadimPushtaev**
https://bugs.python.org/issue34284

### CODE REVIEW

**Comment by pppery** [comment]
Please don't combine patches for two issues in one PR

File: Objects/typeobject.c [hunk 1]
```diff
@@ -5885,12 +5885,24 @@ tp_new_wrapper(PyObject *self, PyObject *args, PyObject *kwds)
         staticbase = staticbase->tp_base;
     /* If staticbase is NULL now, it is a really weird type.
        In the spirit of backwards compatibility (?), just shut up. */
+    char extra[10] = "normal";
```
Comment (pppery) [inline Objects/typeobject.c]: This seems to be dead code

File: Objects/typeobject.c [hunk 2]
```diff
@@ -5885,12 +5885,24 @@ tp_new_wrapper(PyObject *self, PyObject *args, PyObject *kwds)
...
-                     "%s.__new__(%s) is not safe, use %s.__new__()",
-                     type->tp_name,
-                     subtype->tp_name,
-                     staticbase->tp_name);
+        if (subtype->tp_new) {
+            PyErr_Format(PyExc_TypeError,
+                         "%s.__new__(%s) is not safe, use %s.__new__()",
+                         type->tp_name,
+                         subtype->tp_name,
+                         subtype->tp_name);
```
Comment (pppery) [inline Objects/typeobject.c]: See comment on bpo issue

**Comment by VadimPushtaev** [comment]
> Please don't combine patches for two issues in one PR
That was unintentional. Fixed.

**Comment by pppery** [comment]
Also, I disagree with this method of fixing it, as I explained on the BPO issue

File: Lib/test/test_types.py [hunk 3]
```diff
@@ -598,6 +598,18 @@ def test_method_descriptor_types(self):
...
+            object.__new__(dict)
+        message = 'object.__new__(dict) is not safe, use dict.__new__()'
+        self.assertEqual(str(cm.exception), message)
+
+    def test_unsafe_new_no_new(self):
+        with self.assertRaises(TypeError) as cm:
+            object.__new__(type(sys.flags))
+        message = 'object.__new__(sys.flags) is not safe'
+        self.assertEqual(str(cm.exception), message)
+
```
Comment (eric-wieser) [inline Lib/test/test_types.py]: Can you add a test for `tuple.__new__(type(sys.flags))`?

Comment (VadimPushtaev) [reply Lib/test/test_types.py]: That's easy, but I'm still not sure what is the right behavior here. I need time to read through bpo-31506 to comprehend the problem.

**Comment by pppery** [comment]
Discussion on the BPO issue seems to be suggesting a different method of fixing

**Comment by pppery** [comment]
I think that this patch makes the "if(type->tp_new == null)" check in type_call unnecessary

**Comment by pppery** [comment]
This also seems like it won't fix the bug for `type(sys.flags)`, which has its own problem

**Comment by VadimPushtaev** [comment]
Thanks for your comments.

> This also seems like it won't fix the bug for type(sys.flags), which has its own problem

Exactly, there is more than one reason why `tp_new` can be `null`. That's why we can't remove `if(type->tp_new == null)` yet.

I know that there is still work to do. I need to write some tests, check all places where `tp_new` is compared to `null` (there are numerous of them), repair `sys.flags` (and similar objects).

Now I just want to now that I'm following the right path.

File: Misc/NEWS.d/next/Core and Builtins/2018-08-08-20-40-46.bpo-34284.jFqKLF.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+NullNew for old extension types instead of just null
```
Comment (pppery) [inline Misc/NEWS.d/next/Core and Builtins/2018-08-08-20-40-46.bpo-34284.jFqKLF.rst]: This NEWS entry could be worded clearer

**Comment by VadimPushtaev** [comment]
More changes:

* News reworded.
* `sys.flags` is fixed, as well as other tuples of `sys`
* Bunch of necessary tests

**Comment by VadimPushtaev** [comment]
I'm also not sure what to do with all that `tp_new == NULL`. We can't guarantee there are no more types that are written the same way as `sys.flags`. Can we just fix what is initially reported in the issue and leave the rest alone?

**Review by pppery — COMMENTED** [review]
The bug I initially reported would continue to appear for all types whose `tp_new == NULL`, which is why I suggested removing checks for that case: better to fail louder then let a cryptic error-handling bug go unnoticed.  (Although I don't expect such a change to happen) Yes, this change could just happen, but that will still leave the potential for buggy types and make the code (in my opinion -- I expect to be disagreed with here) less clean. Also, the test you added testing `sys.flags` should probably also test the other types you fixed and live in `test_sys`

**Comment by VadimPushtaev** [comment]
> Also, the test you added testing `sys.flags` should probably also test the other types you fixed and live in `test_sys`

Less ambitious tests for this were already there, so it makes perfect sense. Fixed.

**Comment by VadimPushtaev** [comment]
`_curses_panel.panel` aka `curses.panel.panel` is fixed as well.

**Comment by VadimPushtaev** [comment]
Now we can remove that `tp_new == NULL` check. It can be done fairly simple:

```diff
--- a/Objects/typeobject.c
+++ b/Objects/typeobject.c
@@ -912,12 +912,10 @@ type_call(PyTypeObject *type, PyObject *args, PyObject *kwds)
 {
     PyObject *obj;

-    if (type->tp_new == NULL) {
-        PyErr_Format(PyExc_TypeError,
-                     "cannot create '%.100s' instances",
-                     type->tp_name);
-        return NULL;
-    }
+    /* tp_new cannot be NULL,
+       if you want to prevent class instantiation
+       use PyType_NullNew as tp_new instead */
+    assert(type->tp_new != NULL);
```

The Python test suite still passes with that change.

However, in that case, we should update docs as well. The current behavior is documented at least here:

* https://github.com/python/cpython/blame/master/Doc/c-api/typeobj.rst#L1623
* https://github.com/python/cpython/blame/master/Doc/c-api/typeobj.rst#L2422

What about custom user types? Is it OK to break C API between major releases? Sholdn't we depreceate it first?

I need advice.

File: Lib/tkinter/test/test_tkinter/test_internals.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import unittest, test.support
+import _tkinter
+import re
+
+class InternalsTest(unittest.TestCase):
+    def assert_raise_on_new(self, tkinter_type):
+        # Users are intentionally prevented from creating new instances of
+        # sys.flags, sys.version_info, and sys.getwindowsversion.
```
Comment (pppery) [inline Lib/tkinter/test/test_tkinter/test_internals.py]: tkinter =/= sys

**Review by pppery — CHANGES_REQUESTED** [review]
Also, the title of this PR should probably be changed

File: Lib/tkinter/test/test_tkinter/test_internals.py (hunk 5 above)
Comment (VadimPushtaev) [reply Lib/tkinter/test/test_tkinter/test_internals.py]: Fixed.

**Review by pppery — APPROVED** [review]

**Review by asvetlov — CHANGES_REQUESTED** [review]
The PR adds a new *public* `PyType_NullNew` API.
I doubt if we need it, the problem can be solved by with `static` private functions

**Comment by VadimPushtaev** [comment]
@asvetlov, thanks for the review.

First, I'm still looking for an answer for https://github.com/python/cpython/pull/8576#issuecomment-418839461. Should we allow `type->tp_new == NULL`?

Second, if we don't expose `PyType_NullNew` as a public function, we don't allow third-party libraries to make classes with the undefined `__new__`. Is it OK?

**Comment by pppery** [comment]
@asvetlov `tp_new == NULL` is what is currently allowed. This patch is proposing not allowing it

Why should trying to instantiate `_curses_panel.panel` produce a different error message than trying to instantiate any other currently `tp_new == NULL` type, like generator for example? 

Not providing a `PyType_NullNew` function would have the effect of requiring each module that wants to create a heaptype that can't be instantiated by calling it to duplicate its code, a task that seems to have no benefit.

**Comment by asvetlov** [comment]
Sorry, I have to re-read the thread to get restore the PR context first.

**Pull request closed without merging**